processing/src/main/java/org/apache/carbondata/processing/merger/CarbonCompactionExecutor.java

```diff
@@ -140,25 +139,72 @@ public CarbonCompactionExecutor(Map<String, TaskBlockInfo> segmentMapping,
       for (String task : taskBlockListMapping) {
         list = taskBlockInfo.getTableBlockInfoList(task);
-        Collections.sort(list);
-        LOGGER.info(
-            "for task -" + task + "- in segment id -" + segmentId + "- block size is -" + list
-                .size());
-        queryModel.setTableBlockInfos(list);
-        if (sortingRequired) {
-          resultList.get(CarbonCompactionUtil.UNSORTED_IDX).add(
-              new RawResultIterator(executeBlockList(list, segmentId, task, configuration),
-                  sourceSegProperties, destinationSegProperties, false));
-        } else {
-          resultList.get(CarbonCompactionUtil.SORTED_IDX).add(
-              new RawResultIterator(executeBlockList(list, segmentId, task, configuration),
-                  sourceSegProperties, destinationSegProperties, false));
+        // during update there may be a chance that the cardinality may change within the segment
+        // which may lead to failure while converting the row, so get all the blocks present in a
+        // task and then split into multiple lists of same key length and create separate
+        // RawResultIterator for each list of same key length. If all the blocks have same keylength
+        // , then make a single RawResultIterator for all the blocks
+        List<List<TableBlockInfo>> listOfTableBlocksBasedOnKeyLength =
+            getListOfTableBlocksBasedOnKeyLength(list);
+        for (List<TableBlockInfo> tableBlockInfoList : listOfTableBlocksBasedOnKeyLength) {
+          Collections.sort(tableBlockInfoList);
+          LOGGER.info(
+              "for task -" + task + "- in segment id -" + segmentId + "- block size is -" + list
+                  .size());
+          queryModel.setTableBlockInfos(tableBlockInfoList);
+          if (sortingRequired) {
+            resultList.get(CarbonCompactionUtil.UNSORTED_IDX).add(new RawResultIterator(
+                executeBlockList(tableBlockInfoList, segmentId, task, configuration),
+                getSourceSegmentProperties(
+                    Collections.singletonList(tableBlockInfoList.get(0).getDataFileFooter())),
+                destinationSegProperties, false));
+          } else {
+            resultList.get(CarbonCompactionUtil.SORTED_IDX).add(new RawResultIterator(
+                executeBlockList(tableBlockInfoList, segmentId, task, configuration),
+                getSourceSegmentProperties(
+                    Collections.singletonList(tableBlockInfoList.get(0).getDataFileFooter())),
+                destinationSegProperties, false));
+          }
         }
       }
     }
     return resultList;
   }
 
+  /**
+   * This method returns the List of TableBlockInfoList, where each listOfTableBlockInfos will have
+   * same keySize
+   * @param tableBlockInfos List of tableBlock Infos
+   * @return
+   */
+  private List<List<TableBlockInfo>> getListOfTableBlocksBasedOnKeyLength(
```

Comment:
ALL comments inside this method is useless, please consider only to comment with helpful information.

Reply:
updated